### PR TITLE
fix: systemd-networkd reverse route ordering

### PR DESCRIPTION
**Pull request by ssahani**
closes https://github.com/systemd/systemd/issues/5430

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -861,16 +861,31 @@ static int link_enter_set_routes(Link *link) {
...
+        LIST_FOREACH(routes, rt, link->network->static_routes) {
+                if (!in_addr_is_null(rt->family, &rt->gw)) {
+                        r = route_configure(rt, link, route_handler);
+                        if (r < 0) {
+                                log_link_warning_errno(link, r, "Could not set routes: %m");
+                                link_enter_failed(link);
+                                return r;
+                        }
+
+                        link->route_messages++;
```
Comment (yuwata) [inline src/network/networkd-link.c:885]: Double counted?

Comment (ssahani) [reply src/network/networkd-link.c:885]: ooops right thanks .

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -861,14 +861,29 @@ static int link_enter_set_routes(Link *link) {
         link_set_state(link, LINK_STATE_SETTING_ROUTES);
 
         LIST_FOREACH(routes, rt, link->network->static_routes) {
```
Comment (yuwata) [inline src/network/networkd-link.c]: Not a matter, but `{` after `LIST_FOREACH()` can be dropped.

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -861,14 +861,29 @@ static int link_enter_set_routes(Link *link) {
...
+                                log_link_warning_errno(link, r, "Could not set routes: %m");
+                                link_enter_failed(link);
+                                return r;
+                        }
+
+                        link->route_messages++;
                 }
+        }
+
+        LIST_FOREACH(routes, rt, link->network->static_routes) {
```
Comment (yuwata) [inline src/network/networkd-link.c]: Same here.

**Comment by ssahani** [comment]
ya good idea. will update. thanks

**Comment by ssahani** [comment]
updated thanks .

**Comment by yuwata** [comment]
@ssahani Does the test you added pass on your testing environment? The test fails on my laptop...

Yeah, with this change, now `networkd` does not complain "Could not set route: No route to host".
But, `ip route list dev dummy98` shows nothing...

**Comment by ssahani** [comment]
@yuwata  yes 

```
test_routing_policy_rule_port_range (__main__.NetworkdNetWorkTests) ... 0:	from all lookup local 
111:	from 192.168.100.18 tos 0x08 ipproto tcp sport 1123-1150 dport 3224-3290 lookup 7 
1000:	from all lookup [l3mdev-table] 
32766:	from all lookup main 
32767:	from all lookup default
ok
```

but getting failed I will take a look. 
```
test_ip_route_gateway (__main__.NetworkdNetWorkTests) ... 
FAIL

```

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -860,16 +860,29 @@ static int link_enter_set_routes(Link *link) {
 
         link_set_state(link, LINK_STATE_SETTING_ROUTES);
 
-        LIST_FOREACH(routes, rt, link->network->static_routes) {
-                r = route_configure(rt, link, route_handler);
-                if (r < 0) {
-                        log_link_warning_errno(link, r, "Could not set routes: %m");
-                        link_enter_failed(link);
-                        return r;
+        LIST_FOREACH(routes, rt, link->network->static_routes)
```
Comment (poettering) [inline src/network/networkd-link.c:864]: could you insert a comment here saying why the two very similar loops are necessary?

Comment (ssahani) [reply src/network/networkd-link.c:864]: OK updated. thanks for review.

**Comment by yuwata** [comment]
I've confirmed the test passes.

**Pull request merged**